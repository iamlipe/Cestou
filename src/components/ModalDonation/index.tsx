import React, {memo} from 'react';
import styled, {useTheme} from 'styled-components/native';
import {GestureHandlerRootView} from 'react-native-gesture-handler';

import {
  StyledModal,
  StyledBackgroundModal,
  StyledContent,
  StyledContainerButtons as ContainerButtons,
  StyledConfirmButton,
  StyledCancelButton,
  StyledCloseButton,
} from '../Modal';

import Counter from '../Counter';

interface Props {
  name: string;
  control: any;
  error?: string;
  onConfirm: () => any;
  onCancel: () => any;
  onClose: () => any;
}

const ModalDonation = ({
  name,
  control,
  error,
  onConfirm,
  onCancel,
  onClose,
  ...rest
}: Props) => {
  const theme = useTheme();

  return (
    <GestureHandlerRootView>
      <StyledModal animationType="fade" transparent {...rest}>
        <StyledBackgroundModal>
          <StyledContent>
            <StyledCloseButton
              name="close"
              size={20}
              color={theme.colors.GRAY_900}
              backgroundColor={theme.colors.BACKGROUND}
              onPress={onClose}
            />

            <StyledTitle>
              Você escolheu doar para a ONG Criança sem fome!
            </StyledTitle>
            <StyledSubTitle>Você tem 0 moedas</StyledSubTitle>
            <StyledText>Quantas moedas deseja doar?</StyledText>

            <Counter name={name} control={control} />
            {error && <StyledError>{error}</StyledError>}

            <StyledContainerButtons>
              <StyledCancelButton
                buttonColor="text_only"
                textColor="primary"
                size="medium"
                title="Cancelar"
                onPress={onCancel}
              />

              <StyledConfirmButton
                size="medium"
                title="Confirmar"
                onPress={onConfirm}
              />
            </StyledContainerButtons>
          </StyledContent>
        </StyledBackgroundModal>
      </StyledModal>
    </GestureHandlerRootView>
  );
};

const StyledText = styled.Text`
  width: 100%;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  color: ${({theme}) => theme.colors.GRAY_900};
  text-align: center;
  margin-bottom: 16px;
`;

const StyledTitle = styled(StyledText)`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALL};
  color: ${({theme}) => theme.colors.PRIMARY_800};
`;

const StyledSubTitle = styled(StyledText)`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
`;

const StyledContainerButtons = styled(ContainerButtons)`
  margin-top: 24px;
`;

const StyledError = styled.Text`
  align-self: center;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLEST};
  color: ${({theme}) => theme.colors.ERROR_800};
  margin-top: 3px;
`;

export default memo(ModalDonation);
