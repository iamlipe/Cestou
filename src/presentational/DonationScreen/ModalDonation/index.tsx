import React, {memo} from 'react';
import styled, {useTheme} from 'styled-components/native';
import {t} from 'i18next';
import {GestureHandlerRootView} from 'react-native-gesture-handler';

import {
  StyledModal,
  StyledBackgroundModal,
  StyledContent,
  StyledContainerButtons as ContainerButtons,
  StyledConfirmButton,
  StyledCancelButton,
  StyledCloseButton,
} from '@/components/Modal';

import Counter from '@/components/Counter';

interface Props {
  name: string;
  control: any;
  error?: string;
  onConfirm: () => any;
  onCancel: () => any;
  onClose: () => any;
}

const ModalDonation = ({
  name,
  control,
  error,
  onConfirm,
  onCancel,
  onClose,
  ...rest
}: Props) => {
  const theme = useTheme();

  return (
    <GestureHandlerRootView>
      <StyledModal animationType="fade" transparent {...rest}>
        <StyledBackgroundModal>
          <StyledContent>
            <StyledCloseButton
              name="close"
              size={20}
              color={theme.colors.GRAY_900}
              backgroundColor={theme.colors.BACKGROUND}
              onPress={onClose}
            />

            <StyledTitle>
              {t('text.componentModalDonation.titleModalDonation')}
            </StyledTitle>
            <StyledSubTitle>
              {t('text.componentModalDonation.subTitleModalDonation')}
            </StyledSubTitle>
            <StyledText>
              {t('text.componentModalDonation.textModalDonation')}
            </StyledText>

            <Counter name={name} control={control} />
            {error && <StyledError>{error}</StyledError>}

            <StyledContainerButtons>
              <StyledCancelButton
                buttonColor="text_only"
                textColor="primary"
                size="small"
                title={t('button.cancel')}
                onPress={onCancel}
              />

              <StyledConfirmButton
                size="small"
                title={t('button.confirm')}
                onPress={onConfirm}
              />
            </StyledContainerButtons>
          </StyledContent>
        </StyledBackgroundModal>
      </StyledModal>
    </GestureHandlerRootView>
  );
};

const StyledText = styled.Text`
  width: 100%;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  color: ${({theme}) => theme.colors.GRAY_900};
  text-align: center;
  margin-bottom: 16px;
`;

const StyledTitle = styled(StyledText)`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALL};
  color: ${({theme}) => theme.colors.PRIMARY_800};
`;

const StyledSubTitle = styled(StyledText)`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
`;

const StyledContainerButtons = styled(ContainerButtons)`
  margin-top: 24px;
`;

const StyledError = styled.Text`
  align-self: center;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLEST};
  color: ${({theme}) => theme.colors.ERROR_800};
  margin-top: 3px;
`;

export default memo(ModalDonation);
