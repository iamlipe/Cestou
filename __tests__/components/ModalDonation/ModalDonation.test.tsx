import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/Testing';

import ModalDonation from '@/presentational/DonationScreen/ModalDonation';

jest.mock('react-hook-form', () => ({
  useController: () => ({
    field: {
      onChange: jest.fn(),
    },
  }),
}));

describe('ModalDonation', () => {
  const name = 'test';
  const control = jest.fn();
  const title = 'title';
  const subtitle = 'subtitle';
  const text = 'text';
  const onConfirm = jest.fn();
  const onCancel = jest.fn();
  const onClose = jest.fn();

  test('should render correctly', () => {
    renderWithThemeProvider(
      <ModalDonation
        name={name}
        control={control}
        title={title}
        subtitle={subtitle}
        text={text}
        onConfirm={onConfirm}
        onCancel={onCancel}
        onClose={onClose}
      />,
    );
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(
      <ModalDonation
        name={name}
        control={control}
        title={title}
        subtitle={subtitle}
        text={text}
        onConfirm={onConfirm}
        onCancel={onCancel}
        onClose={onClose}
      />,
    );
  });

  test('should render title', () => {
    const modalDonation = renderWithThemeProvider(
      <ModalDonation
        name={name}
        control={control}
        title={title}
        subtitle={subtitle}
        text={text}
        onConfirm={onConfirm}
        onCancel={onCancel}
        onClose={onClose}
      />,
    );

    const titleText = modalDonation.queryByText(title);

    expect(titleText).toBeTruthy();
  });

  test('should render subtitle', () => {
    const modalDonation = renderWithThemeProvider(
      <ModalDonation
        name={name}
        control={control}
        title={title}
        subtitle={subtitle}
        text={text}
        onConfirm={onConfirm}
        onCancel={onCancel}
        onClose={onClose}
      />,
    );

    const subtitleText = modalDonation.queryByText(subtitle);

    expect(subtitleText).toBeTruthy();
  });

  test('should render text', () => {
    const modalDonation = renderWithThemeProvider(
      <ModalDonation
        name={name}
        control={control}
        title={title}
        subtitle={subtitle}
        text={text}
        onConfirm={onConfirm}
        onCancel={onCancel}
        onClose={onClose}
      />,
    );

    const textText = modalDonation.queryByText(text);

    expect(textText).toBeTruthy();
  });

  test('should render error message', () => {
    const modalDonation = renderWithThemeProvider(
      <ModalDonation
        name={name}
        control={control}
        title={title}
        subtitle={subtitle}
        text={text}
        error="someting wrong"
        onConfirm={onConfirm}
        onCancel={onCancel}
        onClose={onClose}
      />,
    );

    const errorText = modalDonation.queryByText('someting wrong');

    expect(errorText).toBeTruthy();
  });
});
