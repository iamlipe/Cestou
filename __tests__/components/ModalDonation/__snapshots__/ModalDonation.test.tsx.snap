// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalDonation should matches snapshot 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    style={
      Array [
        Object {},
      ]
    }
    transparent={true}
    visible={true}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#00000026",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 4,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
              "width": "80%",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "borderRadius": 5,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "padding": 8,
                },
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderRadius": 5,
                },
                Array [
                  Object {
                    "marginRight": -10,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#1A1A1A",
                    "fontSize": 20,
                  },
                  Object {
                    "marginRight": 10,
                  },
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#00843F",
                "fontFamily": "SourceSansPro-SemiBold",
                "fontSize": 31,
                "marginBottom": 16,
                "textAlign": "center",
                "width": "100%",
              },
            ]
          }
        >
          title
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#1A1A1A",
                "fontFamily": "SourceSansPro-SemiBold",
                "fontSize": 27,
                "marginBottom": 16,
                "textAlign": "center",
                "width": "100%",
              },
            ]
          }
        >
          subtitle
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#1A1A1A",
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 27,
                "marginBottom": 16,
                "textAlign": "center",
                "width": "100%",
              },
            ]
          }
        >
          text
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#C1E9CF",
                "borderColor": "#00B761",
                "borderRadius": 16,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flexDirection": "row",
                "height": 32,
                "justifyContent": "space-between",
                "paddingBottom": 0,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 0,
                "width": 86,
              },
            ]
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              onPress={[Function]}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#1A1A1A",
                    "fontSize": 14,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Feather",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              testID="decrement-button"
            >
              
            </Text>
          </View>
          <TextInput
            editable={false}
            style={
              Array [
                Object {
                  "color": "#1A1A1A",
                  "fontFamily": "SourceSansPro-Regular",
                  "fontSize": 27,
                  "paddingBottom": 0,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 0,
                  "width": 10,
                },
              ]
            }
            testID="counter-test"
            value="0"
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              allowFontScaling={false}
              onPress={[Function]}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#1A1A1A",
                    "fontSize": 14,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Feather",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              testID="increment-button"
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 24,
                "width": "80%",
              },
            ]
          }
        >
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "transparent",
                "borderRadius": 4,
                "flexDirection": "row",
                "height": 38,
                "justifyContent": "center",
                "marginBottom": 10,
                "opacity": 1,
                "width": 90,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#00843F",
                    "fontFamily": "SourceSansPro-SemiBold",
                    "fontSize": 27,
                  },
                ]
              }
              textColor="primary"
            >
              Cancelar
            </Text>
          </View>
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#00843F",
                "borderRadius": 4,
                "flexDirection": "row",
                "height": 38,
                "justifyContent": "center",
                "marginBottom": 10,
                "opacity": 1,
                "width": 90,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "SourceSansPro-SemiBold",
                    "fontSize": 27,
                  },
                ]
              }
              textColor="white"
            >
              Confirmar seleção
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
